[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81853

These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

Commit:
[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives

These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:
```diff
@@ -226,170 +225,13 @@ def _result_distribute_with_col_rearrange(
     return output.index_select(0, torch.tensor(indices, device=output.device))
 
 
-def _handle_row_wise_lookup_distribute(
```
does it mean that rowwise look up is not needed anymore?

fduwjj replied:
Yeah this is old logic. We don't need this anymore so that we can get rid of unnecessary collective communications.

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:
```diff
@@ -39,34 +41,25 @@ def sharded_embedding(types, args, kwargs, pg):
 
-    1. First the input is flattened to 1D and gets sorted so that we can distribute
-       them to the corresponding rank. For example if the given input is
+    1. First the input is broadcasted to all ranks, since this is SPMD we
```
should this be an all_gather or broadcast, looks confusing to me, usually if input is sharded, we all_gather, if input is not, we broadcast it to get the same values, which one we are taking here?

wanchaol reviewed (commented):
Thanks for refactoring this! I did an initial pass, it looks great, have a few comments on when to use all_gather, let me know if those make sense.

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:
```diff
@@ -39,34 +41,25 @@ def sharded_embedding(types, args, kwargs, pg):
+       For example on rank 0, we store ID [0, 1, 2]. We only keep the ID
+       inside the set of numbers. The rest of them will be masked to an extra row.
+       The masked matrix will be used for embedding look up and is like:
+       tensor([[4, 4, 2, 4, 4, 4],
```
4 is the masked ID and the reason is that for the weight being sharded there is only 0, 1 and 2. We will add one more dummy entry (4); otherwise, when max_norm is specified, the result is not correct because max_norm will norm the row which is looked up.

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:
```diff
@@ -443,4 +283,71 @@ def _handle_max_norm_col_wise(
+
+    # When max_norm is set, it will only re-norm the row being looked up.
+    padding_row = torch.zeros(
+        1, weight.size(1), device=gather_inp.device, dtype=weight.dtype
```
No, this is the extra row we padded to the weight and weight is a sharded tensor it does not have device implemented.

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:
```diff
@@ -215,11 +206,11 @@ def _handle_col_wise_sharding(
     Returns: final result of lookup.
     """
-    gathered_inputs = None
+    gathered_inputs = all_gather(input, group=pg)
```
Oh, sure. Let me add that.

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding_bag.py:
```diff
@@ -377,443 +352,132 @@ def _handle_row_wise_sharding(
+                gathered_per_sample_weights,
+                gathered_offsets,
+            ) = _all_gather_embedding_bag_input(input, per_sample_weights, offsets, pg)
+            cat_dim = 0 if input.dim() > 1 else -1
```
This is for differentiating the case when dim == 1 and dim > 1. Dim will not < 1.

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:
```diff
@@ -406,24 +248,22 @@ def _handle_max_norm_col_wise(
-    # allgather the inputs first.
-    gathered_inputs = [torch.zeros_like(input) for _ in range(world_size)]
-    dist.all_gather(gathered_inputs, input, group=pg)
     unique_inp = torch.unique(torch.cat(gathered_inputs))
```
Let me think about this..

fduwjj replied:
After a second thought, we probably need to keep using all_gather for col-wise sharding because we need to make it a list and do all the rearrangement. However, when we migrating to DT, this should not be a problem at all.

Commit:
Update on "[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives"


These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good! please fix the lint and make sure multigpu jobs pass before landing (add a ciflow/periodic to trigger it)

Commit:
Update on "[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives"


These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

[ghstack-poisoned]

fduwjj commented:
@pytorchmergebot merge this please

fduwjj commented:
@pytorchmergebot merge

Commit:
Update on "[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives"


These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

[ghstack-poisoned]

fduwjj commented:
@pytorchmergebot merge
